File: 01.hs
import Data.Char
import Data.List
import Data.List.Utils
import Data.Maybe
import Data.String.Utils

import Common

main = do
    content <- readFile "01.input"
    let ls = lines content
    putStrLn $ show $ ls

    -- let s = sum $ map findDigits ls
    -- putStrLn $ show $ s
    -- putStrLn $ show $ map (\x -> (x, replaceDigits x, findDigits $ replaceDigits x)) ls

    let s_ = sum $ map findDigits $ ls
    putStrLn $ show $ s_

findDigit digits cs
    | c >= '0' && c <= '9' = (ord c) - (ord '0')
    | isJust f = snd $ fromJust f
    | otherwise = findDigit digits $ tail cs
    where
        c = head cs
        f = find (\(x, y) -> startswith x cs) digits

findDigits l = d1 * 10 + d2
    where
        digits = [("one", 1), ("two", 2), ("three", 3), ("four", 4), ("five", 5), ("six", 6), ("seven", 7), ("eight", 8), ("nine", 9)]
        rdigits = map (\(x, y) -> (reverse x, y)) digits
        d1 = findDigit digits l
        d2 = findDigit rdigits $ reverse l

findDigits_ l = d1 * 10 + d2
    where
        digits = filter (\c -> c >= '0' && c <= '9') l
        d1 = (ord $ head digits) - (ord '0')
        d2 = (ord $ last digits) - (ord '0')
    
replaceDigits l =    
    foldl (\s (a, b) -> replace a b s) l [("one", "1"), ("two", "2"), ("three", "3"), ("four", "4"), ("five", "5"), ("six", "6"), ("seven", "7"), ("eight", "8"), ("nine", "9")]

File: 01.ml
open Common
open Printf


let rec find_digit digits_map line : int =
    let c = String.get line 0 in
    if c >= '0' && c <= '9' then
        int_of_char c
    else
        let m = List.find_opt (fun (name, _) -> String.starts_with ~prefix:name line) digits_map in
        if Option.is_some m then
            snd @@ Option.get m
        else
            find_digit digits_map @@ Str.string_after line 1

let find_digits digits_map line : int =
    let reverse_digits_map = List.map (on_fst reverse_string) digits_map in
    let d1 = find_digit digits_map line in
    let d2 = find_digit reverse_digits_map @@ reverse_string line in
    d1 * 10 + d2


let lines = read_lines @@ input_file

let ds = sum @@ List.map (find_digits []) lines
let () = printf "%d\n" ds

let digits = [("one", 1); ("two", 2); ("three", 3); ("four", 4); ("five", 5); ("six", 6); ("seven", 7); ("eight", 8); ("nine", 9)]
let ds = sum @@ List.map (find_digits digits) lines
let () = printf "%d\n" ds
